File: CartQuestiOS/CartQuestiOS/Models/KrogerProduct.swift
import Foundation

struct KrogerProduct: Codable, Identifiable {
    var id: String { productId }
    var productId: String
    var description: String
    var brand: String?
    var images: [KrogerImage]?
    var items: [KrogerItemPrice]?

    var bestImageUrl: String? {
        let featured = images?.first(where: { $0.featured == true })
        return (featured ?? images?.first)?.bestUrl
    }
}

struct KrogerImage: Codable {
    var perspective: String?
    var sizes: [KrogerImageSize]?
    var featured: Bool?

    private static let sizePreference = ["xlarge", "large", "medium", "small", "thumbnail"]

    var bestUrl: String? {
        guard let sizes else { return nil }
        for preferred in Self.sizePreference {
            if let match = sizes.first(where: { $0.size == preferred }) {
                return match.url
            }
        }
        return sizes.last?.url
    }
}

struct KrogerImageSize: Codable {
    var size: String
    var url: String
}

struct KrogerItemPrice: Codable {
    var price: KrogerPrice?
    var fulfillment: KrogerFulfillment?
}

struct KrogerPrice: Codable {
    var regular: Double?
    var promo: Double?
}

struct KrogerFulfillment: Codable {
    var inStore: Bool
}

File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/data/model/KrogerProduct.kt
package com.amoghghadge.cartquestandroid.data.model

data class KrogerProduct(
    val productId: String,
    val description: String,
    val brand: String? = null,
    val images: List<KrogerImage>? = null,
    val items: List<KrogerItemPrice>? = null
) {
    val bestImageUrl: String?
        get() {
            val featured = images?.firstOrNull { it.featured == true }
            return (featured ?: images?.firstOrNull())?.bestUrl
        }
}

data class KrogerImage(
    val perspective: String? = null,
    val sizes: List<KrogerImageSize>? = null,
    val featured: Boolean? = null
) {
    val bestUrl: String?
        get() {
            val prefs = listOf("xlarge", "large", "medium", "small", "thumbnail")
            for (preferred in prefs) {
                val match = sizes?.firstOrNull { it.size == preferred }
                if (match != null) return match.url
            }
            return sizes?.lastOrNull()?.url
        }
}

data class KrogerImageSize(
    val size: String,
    val url: String
)

data class KrogerItemPrice(
    val price: KrogerPrice?,
    val fulfillment: KrogerFulfillment?
)

data class KrogerPrice(
    val regular: Double? = null,
    val promo: Double? = null
)

data class KrogerFulfillment(
    val inStore: Boolean
)
